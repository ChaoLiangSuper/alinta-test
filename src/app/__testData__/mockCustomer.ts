export default {
  key: '111111',
  firstName: 'first',
  lastName: 'last',
  dateOfBirth: '10/27/2018',
};
