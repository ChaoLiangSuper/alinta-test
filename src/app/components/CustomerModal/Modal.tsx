import styled from 'styled-components';

const Modal = styled.div`
  width: 30%;
  display: flex;
  background: white;
  box-shadow: 0 0 20px lightgray;
  border-radius: ${({ theme }) => theme.borderRadius(3)};
  padding: ${({ theme }) => theme.spacing(5)};
  flex-direction: column;
`;

export default Modal;
