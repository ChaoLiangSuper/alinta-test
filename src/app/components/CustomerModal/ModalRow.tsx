import styled from 'styled-components';

interface ModalRowProps {
  center?: boolean;
}

const ModalRow = styled.div<ModalRowProps>`
  width: 100%;
  display: flex;
  justify-content: ${({ center }) => (center ? 'center' : '')};

  & + & {
    margin-top: ${({ theme }) => theme.spacing()};
  }
`;

export default ModalRow;
