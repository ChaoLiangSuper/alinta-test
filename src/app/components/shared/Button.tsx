import styled from 'styled-components';

interface ButtonProps {
  primary?: boolean;
}

const StyledButton = styled.button<ButtonProps>`
  border: none;
  padding: ${({ theme }) => theme.spacing()};
  border-radius: ${({ theme }) => theme.borderRadius()};
  background: ${({ theme, primary }) =>
    primary ? theme.primaryColor : theme.secondaryColor};
  color: ${({ theme, primary }) =>
    primary ? theme.primaryText : theme.secondaryText};

  &:active {
    background-color: ${({ theme, primary }) =>
      primary ? theme.primaryColorDark : theme.secondaryColorDark};
  }

  & + & {
    margin-left: ${({ theme }) => theme.spacing()};
  }
`;

export default StyledButton;
