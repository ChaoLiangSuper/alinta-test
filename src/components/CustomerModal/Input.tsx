import React from 'react';
import styled from 'styled-components';

interface StyledInputProps {
  invalid?: boolean;
}

interface InputProps extends StyledInputProps {
  name: string;
  label: string;
  date?: boolean;
  value: string;
  onChange: (value: string) => void;
}

const StyledLabel = styled.label`
  display: flex;
  width: 100%;
`;

const StyledText = styled.span`
  flex-grow: 2;
  flex-basis: 0;
`;

const StyledInput = styled.input<StyledInputProps>`
  flex-grow: 3;
  flex-basis: 0;
  width: 100%;
  border: ${({ invalid }) => (invalid ? '1px solid red' : '1px solid grey')};
  border-radius: ${({ theme }) => theme.borderRadius()};
`;

const Input: React.FC<InputProps> = ({
  name,
  label,
  date,
  value,
  onChange,
  invalid,
}) => {
  return (
    <StyledLabel>
      <StyledText>{label}</StyledText>
      <StyledInput
        autoComplete='off'
        name={name}
        type={date ? 'date' : 'text'}
        value={value}
        invalid={invalid}
        onChange={({ target }) => onChange(target.value)}
      />
    </StyledLabel>
  );
};

export default Input;
