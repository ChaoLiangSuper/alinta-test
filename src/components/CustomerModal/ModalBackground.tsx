import styled from 'styled-components';

const ModalBackground = styled.div`
  background: transparent;
  position: absolute;
  width: 100%;
  height: 100%;
  z-index: 100;
  top: 0px;
  left: 0px;
  display: flex;
  justify-content: center;
  align-items: center;
`;

export default ModalBackground;
