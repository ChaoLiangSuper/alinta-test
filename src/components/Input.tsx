import React from 'react';
import styled from 'styled-components';

interface StyledInputProp {
  invalid?: boolean;
}

interface InputProp extends StyledInputProp {
  name: string;
  label: string;
  date?: boolean;
  value: string;
  onChange: (value: string) => void;
}

const StyledLabel = styled.label`
  display: flex;
  width: 100%;
`;

const StyledText = styled.span`
  flex-grow: 2;
  flex-basis: 0;
`;

const StyledInput = styled.input<StyledInputProp>`
  flex-grow: 3;
  flex-basis: 0;
  width: 100%;
  border: ${({ invalid }) => (invalid ? '1px solid red' : '1px solid grey')};
  border-radius: ${({ theme }) => theme.borderRadius()};
`;

const Input: React.FC<InputProp> = ({
  name,
  label,
  date,
  value,
  onChange,
  invalid,
}) => {
  return (
    <StyledLabel>
      <StyledText>{label}</StyledText>
      <StyledInput
        name={name}
        type={date ? 'date' : 'text'}
        value={value}
        invalid={invalid}
        onChange={({ target }) => onChange(target.value)}
      />
    </StyledLabel>
  );
};

export default Input;
