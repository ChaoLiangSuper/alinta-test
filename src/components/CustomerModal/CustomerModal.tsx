import _ from 'lodash';
import React from 'react';
import { connect } from 'react-redux';
import Button from '../shared/Button';
import Input from '../shared/Input';
import { Field, Customer, Store } from '../../../types';
import ModalBackground from './ModalBackground';
import Modal from './Modal';
import ModalRow from './ModalRow';

interface ModalProps {
  onSave: (value: Customer) => void;
  onClose: () => void;
  initialData?: Customer;
  selectedCustomerKey: string | null;
}

interface ModalState {
  value: Customer;
  invalidFields: string[];
}

const fields: Field[] = [
  { label: 'First Name', name: 'firstName' },
  {
    label: 'Last Name',
    name: 'lastName',
  },
  {
    label: 'Date of Birth',
    name: 'dateOfBirth',
    type: 'date',
  },
];

class CustomerModal extends React.Component<ModalProps, ModalState> {
  state: ModalState = {
    value: this.props.initialData
      ? this.props.initialData
      : {
          firstName: '',
          lastName: '',
          dateOfBirth: '',
        },
    invalidFields: [],
  };

  handleChange = (key: string) => (value: string) => {
    this.setState({
      value: {
        ...this.state.value,
        [key]: value,
      },
    });
  };

  handleSubmit = () => {
    const { value } = this.state;
    const { onSave, onClose } = this.props;

    const invalidFields: string[] = [];
    _.forEach(fields, ({ name }) => {
      if (_.isEmpty(value[name].trim())) {
        invalidFields.push(name);
      }
    });

    if (!_.isEmpty(invalidFields)) {
      return this.setState({
        invalidFields,
      });
    }

    onSave(value);
    onClose();
  };

  render() {
    const { onClose } = this.props;
    const { invalidFields, value } = this.state;

    return (
      <ModalBackground>
        <Modal>
          {_.map(fields, ({ name, label, type }) => (
            <ModalRow key={name}>
              <Input
                name={name}
                label={label}
                value={value[name]}
                invalid={invalidFields.indexOf(name) !== -1}
                onChange={this.handleChange(name)}
                date={type === 'date'}
              />
            </ModalRow>
          ))}
          <ModalRow center>
            <Button onClick={onClose}>Cancel</Button>
            <Button primary onClick={this.handleSubmit}>
              Save
            </Button>
          </ModalRow>
        </Modal>
      </ModalBackground>
    );
  }
}

export default connect(
  ({ customers }: Store, { selectedCustomerKey }: ModalProps) =>
    selectedCustomerKey
      ? {
          initialData: customers[selectedCustomerKey],
        }
      : {}
)(CustomerModal);
