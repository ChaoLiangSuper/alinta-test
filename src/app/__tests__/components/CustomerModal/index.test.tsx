import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import Button from '../../../components/shared/Button';
import { CustomerModal as Component } from '../../../components/CustomerModal/CustomerModal';
import mockCustomer from '../../../__testData__/mockCustomer';
import Input from '../../../components/shared/Input';

describe('<CustomerModal />', () => {
  let wrapper: ShallowWrapper;
  const mockFunc = jest.fn();

  beforeEach(() => {
    wrapper = shallow(
      <Component
        onSave={mockFunc}
        onClose={mockFunc}
        selectedCustomerKey={null}
      />
    );
  });

  it('should render without error', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should be able to update value', () => {
    const inputs = wrapper.find(Input);
    expect(inputs.length).toBe(3);
    inputs.first().prop('onChange')('first-test');
    inputs.at(1).prop('onChange')('last-test');
    inputs.last().prop('onChange')('10/10/1010');
    expect(wrapper.state()).toEqual({
      invalidFields: [],
      value: {
        dateOfBirth: '10/10/1010',
        firstName: 'first-test',
        lastName: 'last-test',
      },
    });
  });

  it('should be able to validate inputs', () => {
    wrapper
      .find(Button)
      .last()
      .simulate('click');
    expect(wrapper.state()).toEqual({
      invalidFields: ['firstName', 'lastName', 'dateOfBirth'],
      value: { dateOfBirth: '', firstName: '', lastName: '' },
    });
  });

  it('should be able to save data if value is valid', () => {
    const value = {
      dateOfBirth: '10/10/1010',
      firstName: 'first-test',
      lastName: 'last-test',
    };
    wrapper.setState({
      value,
    });
    wrapper
      .find(Button)
      .last()
      .simulate('click');
    expect(mockFunc).toHaveBeenCalledTimes(2);
    expect(mockFunc).toHaveBeenCalledWith(value);
  });

  it('should be able to pre-fill input with initialData', () => {
    wrapper = shallow(
      <Component
        onSave={mockFunc}
        onClose={mockFunc}
        selectedCustomerKey={null}
        initialData={mockCustomer}
      />
    );
    const inputs = wrapper.find(Input);
    expect(inputs.first().prop('value')).toBe('first');
    expect(inputs.at(1).prop('value')).toBe('last');
    expect(inputs.last().prop('value')).toBe('10/27/2018');
  });
});
