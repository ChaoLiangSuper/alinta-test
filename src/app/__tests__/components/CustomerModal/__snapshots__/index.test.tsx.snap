// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CustomerModal /> should render without error 1`] = `
<styled.div>
  <styled.div>
    <styled.div
      key="firstName"
    >
      <Input
        date={false}
        invalid={false}
        label="First Name"
        name="firstName"
        onChange={[Function]}
        value=""
      />
    </styled.div>
    <styled.div
      key="lastName"
    >
      <Input
        date={false}
        invalid={false}
        label="Last Name"
        name="lastName"
        onChange={[Function]}
        value=""
      />
    </styled.div>
    <styled.div
      key="dateOfBirth"
    >
      <Input
        date={true}
        invalid={false}
        label="Date of Birth"
        name="dateOfBirth"
        onChange={[Function]}
        value=""
      />
    </styled.div>
    <styled.div
      center={true}
    >
      <styled.button
        onClick={[MockFunction]}
      >
        Cancel
      </styled.button>
      <styled.button
        onClick={[Function]}
        primary={true}
      >
        Save
      </styled.button>
    </styled.div>
  </styled.div>
</styled.div>
`;
